core/src/main/scala/org/apache/spark/sql/internal/session.scala

```diff
@@ -565,6 +580,23 @@ private[sql] final class PreprocessTable(state: SnappySessionState) extends Rule
               "generates the same number of columns as its schema.")
       }
 
+    case i@InsertIntoTable(l@LogicalRelation(r: MutableRelation,
+    _, _), _, child, overwrite, _) if l.resolved && child.resolved && overwrite.enabled => {
+        EliminateSubqueryAliases(child) match {
+          case plan: LogicalPlan =>
+            val srcRelationNames = plan.collect {
+              case LogicalRelation(src: MutableRelation, _, _) => src.table.toUpperCase
+            }
+            // we store table as Uppercase in relation
+            if (srcRelationNames.contains(r.table.toUpperCase)) {
+              throw new AnalysisException(
+                s"Cannot overwrite table ${r.table} that is also being read from")
+            }
+          case _ =>
+        }
+      i
+    }
```

Comment:
Why would the rule in PreWriteCheck for InsertIntoTable not apply? If due to table upper case issue then the equals of JDBCMutableRelation and JDBCAppendableRelation can be overridden.